### PR TITLE
Fix to data speed for end flag

### PR DESCRIPTION
**Pull request by alan-stokes**
fixes both end flag and loss of the first packet.

### CODE REVIEW

**Comment by alan-stokes** [comment]
requires 
https://github.com/SpiNNakerManchester/SpiNNMachine/pull/57

to be merged first.

**Comment by Christian-B** [comment]
Run:
https://github.com/SpiNNakerManchester/sPyNNaker8/blob/master/p8_integration_tests/test_0_1_time_steps/test_v/test_synfire_0dot1_timestep_test_get_v.py

With all other repros on master so including https://github.com/SpiNNakerManchester/SpiNNMachine/pull/57.

E           SpinnmanGenericProcessException: 
E                Received exception class: SpinnmanUnexpectedResponseCodeException 
E                With message: Unexpected response RC_ROUTE while performing operation RouterInit using command CMD_RTR 
E                When sending to 7:3:0
E                Stack trace: None

**Comment by alan-stokes** [comment]
i call fibs. try again

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py [hunk 1]
```diff
@@ -626,7 +589,7 @@ def _print_missing(seq_nums):
         last_seq_num = 0
         for seq_num in sorted(seq_nums):
             if seq_num != last_seq_num + 1:
-                print "from list i'm missing sequence num {}".format(seq_num)
+                print "from list i'm missing sequence num {}\n".format(seq_num)
```
Comment (rowleya) [inline spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: Not sure you want the \n in a python print (python prints \n anyway)

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py [hunk 2]
```diff
@@ -358,7 +372,7 @@ def _determine_and_retransmit_missing_seq_nums(
         # locate missing sequence numbers from pile
         missing_seq_nums = self._calculate_missing_seq_nums(seq_nums)
         lost_seq_nums.append(len(missing_seq_nums))
-        # self._print_missing(seq_nums)
+        self._print_missing(seq_nums)
```
Comment (rowleya) [inline spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: Should this be disabled for the live code?

File: c_common/models/extra_monitor_support/extra_monitor_support.c [hunk 3]
```diff
@@ -721,22 +724,22 @@ void read(uint32_t dma_tag, uint32_t offset, uint32_t items_to_read) {
...
 
 //! \brief dma complete callback for reading for original transmission
 void dma_complete_reading_for_original_transmission(){
     //do dma
     uint32_t current_dma_pointer = transmit_dma_pointer;
-    uint32_t key_to_transmit = key_to_transmit_with;
+    uint32_t key_to_transmit = basic_data_key;
     uint32_t items_read_this_time = num_items_read;
 
     // put size in bytes if first send
```
Comment (rowleya) [inline c_common/models/extra_monitor_support/extra_monitor_support.c]: Fix this comment ;)

File: c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c [hunk 4]
```diff
@@ -51,6 +51,7 @@ static uint32_t infinite_run;
 //! the key that causes sequence number to be processed
 static uint32_t new_sequence_key = 0;
 static uint32_t first_data_key = 0;
+static uint32_t end_flag_key = 0;
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c]: agree

File: c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c [hunk 5]
```diff
@@ -119,11 +120,13 @@ void receive_data(uint key, uint payload) {
...
-            if (position_in_store == 2) {
-                data[0] = 0xFFFFFFFF;
-                position_in_store = 1;
-            }
+        if (key == end_flag_key){
+            // set end flag bit in seq num
+            data[0] = data[0] + (1 << 31);
+
+            // adjust size as last payload not counted
+            position_in_store = position_in_store - 1;
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c]: this is likely due to the fact that theres a impicit else, which is where your adding to the pile, but its not big enough to send yet. 

if you increment in a explicit version fo the else, youd also need to do so at the when ready, and change the if to position < items -1

i think thast more clunky than just going back 1 when your dealing with a end key.

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py [hunk 6]
```diff
@@ -72,9 +72,17 @@ class DataSpeedUpPacketGatherMachineVertex(
     # what governs a end flag
     END_FLAG = 0xFFFFFFFF
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: agree

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py (hunk 2 above)
Comment (alan-stokes) [reply spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: yes, was used for debug with all the faff of the last few days.

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py [hunk 7]
```diff
@@ -468,92 +478,45 @@ def _process_data(self, data, first, seq_num, seq_nums, finished,
...
-                    finished = self._determine_and_retransmit_missing_seq_nums(
-                        placement=placement, transceiver=transceiver,
-                        seq_nums=seq_nums, lost_seq_nums=lost_seq_nums)
+        first_packet_element = struct.unpack_from(
+            "<I", data, 0)[0]
+
+        # get flags
+        seq_num = first_packet_element & 0x7FFFFFFF
+        is_end_of_stream = \
+            True if (first_packet_element >> 31) & 1 == 1 else False
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: if made a named variable such as LAST_MESSAGE_FLAG_BIT_MASK then i concur

File: spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py (hunk 1 above)
Comment (alan-stokes) [reply spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py]: was during a debug as wasnt seeing what i expected. later on found it was somethign very differnet. removed

File: c_common/models/extra_monitor_support/extra_monitor_support.c (hunk 3 above)
Comment (alan-stokes) [reply c_common/models/extra_monitor_support/extra_monitor_support.c]: done

File: c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (rowleya) [reply c_common/models/data_speed_up_packet_gatherer/data_speed_up_packet_gatherer.c]: Sorry missed that it was an else if.  Retracted

**Pull request merged**